scripts/metadata_enrichment/tob_bioheart_sequencing_dates.py

```diff
@@ -0,0 +1,234 @@
+        )
+        return defaultdict()
+
+
+# TODO: combine pd.apply() into a single expression
+def extract_excel():
+    """
+    Reads in metadata stored in excel spreadsheets. Appends these into
+    a single dataframe and performs required transformations to columns
+    to extract the desired data.
+
+    :return: Fluid tube (key) IDS mapped to sequencing dates (value)
+    :rtype: dict
+    """
+    tob_workbook_names = [
+        'scripts/metadata_enrichment/1K1K Sequencing Dates.xlsx',
+        'scripts/metadata_enrichment/BioHEART Sequencing Dates.xlsx',
+    ]
+    df_list = []
+
+    # Amalgamate data in all sheets listed in tob_sheet_names and bioheart_sheet_names
+    for workbook in tob_workbook_names:
+        df_sheet_index = pd.read_excel(workbook, sheet_name=None, header=0)
+        temp_df = pd.concat(df_sheet_index)
+        df_list.append(temp_df)
+
+    aggregated_df = pd.concat(df_list, ignore_index=True)
+
+    # Separate accession date and fluidX_id in Sample Identifier column
+    aggregated_df['accession_date'] = aggregated_df['Sample Identifier'].apply(
+        lambda x: x.split('_')[0],
+    )
+    aggregated_df['fluidX_id'] = aggregated_df['Sample Identifier'].apply(
+        lambda x: x.split('_')[1],
+    )
+
+    # Insert accession value into new dictionary fluidX_to_sequencing_date. Key is FluidX ID
+    return pd.Series(
+        aggregated_df.accession_date.values,
+        index=aggregated_df.fluidX_id,
+    ).to_dict()
+
+
+# TODO: combine api calls and
+async def upsert_sequencing_dates(
+    fluidx_to_assay_ids: defaultdict,
+    fluidx_to_sequencing_date: dict,
+):
+    # construct API update calls
+    # Iterate through the fluidX_to_assay_IDs dictionary because this is representative of what's already in metamist
+    # That is: fluidX_to_assay_IDs groups assays by fluidX_tube_id
+
+    assay_api = AssayApi()
+    assays_to_update = []
+
+    for fluidx_id, assay_ids in fluidx_to_assay_ids.items():
+        # Get sequencing date for each fluidX id from fluidX_to_sequencing_date dict
+        if fluidx_id:
+            try:
+                # Create object required to update assay
+                date = fluidx_to_sequencing_date[fluidx_id]
+            except KeyError:
+                print(f'Tube ID {fluidx_id} is not in the sequencing manifest')
+            else:
+                for assay_id in assay_ids:
+                    updated_assay = AssayUpsert()
+                    updated_assay['id'] = assay_id
+                    updated_assay['meta'] = {}
+                    updated_assay['meta']['fluid_x_tube_sequencing_date'] = date
+                    assays_to_update.append(updated_assay)
+                    print(f'API call added for {assay_id}')
+
+        else:
+            print(f'*****\nNO TUBE ID FOR : {fluidx_id} and assays {assay_ids}')
+
+    return await assay_api.upsert_assay_async(assay_upsert=list(assays_to_update))
```

Comment:
I dont think this way of calling the API endpoint will work. This endpoint is expecting a single AssayUpsert object as input, so it would need to go inside the for loop above. This is definitely one that we would want to `gather` a number of individual async calls. Let's chat about it tomorrow.